node/lib/statsd.js

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var assert = require('assert');
+
+function TChannelStatsd(channel, statsd) {
+    if (!(this instanceof TChannelStatsd)) {
+        return new TChannelStatsd(channel, statsd);
+    }
+
+    var self = this;
+    self.statsd = statsd;
+    self.channel = channel;
+    self.channel.on('stat', onStat);
+
+    function onStat(stat) {
+        self.onStat(stat);
+    }
+}
+
+function clean(str) {
+    return str
+        .replace(/:/g, '-')
+        .replace(/\./g, '-')
+        .replace(/{|}/g, '-');
+}
+
+function getKey(common, stat) {
+    var prefix = ['tchannel', stat.name].join('.');
+    switch (stat.name) {
+        // outbound
+        case 'outbound.calls.sent':
+            return [
+                prefix,
+                clean(stat.tags.service),
+                clean(stat.tags['target-service']),
+                clean(stat.tags['target-endpoint'])
+            ].join('.');
```

Comment:
At some point we figured out that [].join() style interpolation is actually slower than string concatenation in V8.